File: lib/preferences.rb
# typed: strict
require 'sorbet-runtime'
require './lib/person'
require 'pry-byebug'

class Preferences

  extend T::Sig

  sig {params(mentees: T::Array[Person2025], mentors: T::Array[Person2025]).returns(T::Hash[Person2025, T::Array[Person2025]])} 
  def self.compute_mentor_to_mentee_preferences(mentees:, mentors:)
    preferences = T.let({}, T::Hash[Person2025, T::Array[Person2025]])

    if !mentees.all?(&:is_mentee)
      raise "Found non mentees!"
    end

    if !mentors.all?(&:is_mentor)
      raise "Found non mentors!"
    end

    mentors.each do |mentor|

      if !mentor.is_mentor
        puts "Filtering out non-mentor #{mentor.name}"
        preferences[mentor] = []
        next
      end

      potential_mentees = mentees
        # Only keep people who want to be mentees.
        .select {|other_person| other_person.is_mentee}
        # Rule out yourself as a potential mentor.
        .reject {|other_person| mentor == other_person}
        # Only keep potential mentees who are more junior.
        .select {|other_person| other_person.seniority < mentor.seniority}

      if potential_mentees.size == 0
        puts "Could not find any potential mentees for #{mentor.name}"
      end

      # Perform a cascading comparison where we sort based on adjacent seniority, city, state, then region
      # (in descending order of priority).
      preferred_mentees = potential_mentees.sort do |p1, p2|
        comparisons = [
          compare_international_preference_for_mentor(mentor: mentor, p1: p1, p2: p2),
          compare_mentee_seniority_allowlist(mentor: mentor, p1: p1, p2: p2),
          compare_preferring_target(target: mentor.state, a: p1.state, b: p2.state),
          compare_rank(target_rank: mentor.seniority, p1_rank: p1.seniority, p2_rank: p2.seniority),
        ]

        final_comparison = sort_by_comparison_list(comparisons)

        # We want to sort in descending order (aka most preferred mentor first).
        -1 * final_comparison
      end

      preferences[mentor] = preferred_mentees
    end

    preferences
  end

  sig {params(mentees: T::Array[Person2025], mentors: T::Array[Person2025]).returns(T::Hash[Person2025, T::Array[Person2025]])}
  def self.compute_mentee_to_mentor_preferences(mentees:, mentors:)
    preferences = T.let({}, T::Hash[Person2025, T::Array[Person2025]])

    if !mentees.all?(&:is_mentee)
      raise "Found non mentees!"
    end

    if !mentors.all?(&:is_mentor)
      raise "Found non mentors!"
    end

    mentees.each do |mentee|
      potential_mentors = mentors
        # Rule out yourself as a potential mentor.
        .reject {|other_person| mentee == other_person}
        # Only keep potential mentors who are more senior
        .select {|other_person| other_person.seniority > mentee.seniority}

      # Perform a cascading comparison where we sort (from most to least important):
      # - difference in seniority (prefer closeness)
      # - state (prefer closeness)
      preferred_mentors = potential_mentors.sort do |p1, p2|
        comparisons = [
          compare_international_preference_for_mentee(mentee: mentee, p1: p1, p2: p2),
          compare_preferring_target(target: mentee.state, a: p1.state, b: p2.state),
          compare_rank(target_rank: mentee.seniority, p1_rank: p1.seniority, p2_rank: p2.seniority),
        ]

        final_comparison = sort_by_comparison_list(comparisons)

        # We want to sort in descending order (aka most preferred mentor first).
        -1 * final_comparison
      end

      preferences[mentee] = preferred_mentors
    end

    preferences
  end

  # Performing a cascading comparison where we return the first non-zero value (aka
  # the tie-breaker), or 0 if there is none.
  sig {params(comparisons: T::Array[Integer]).returns(Integer)}
  private_class_method def self.sort_by_comparison_list(comparisons)
    comparisons.each_with_index do |comparison, idx|
      if comparison != 0
        return comparison
      end
    end
    0
  end

    sig do
    params(
      mentee: Person2025,
      p1: Person2025,
      p2: Person2025
    ).returns(Integer)
  end
  private_class_method def self.compare_international_preference_for_mentee(mentee:, p1:, p2:)
    if !mentee.is_international
      return 0
    end

    if p1.prefers_mentoring_international && !p2.prefers_mentoring_international
      return 1
    end

    if !p1.prefers_mentoring_international && p2.prefers_mentoring_international
      return -1
    end

    return 0
  end

  sig do
    params(
      mentor: Person2025,
      p1: Person2025,
      p2: Person2025
    ).returns(Integer)
  end
  private_class_method def self.compare_international_preference_for_mentor(mentor:, p1:, p2:)
    if !mentor.prefers_mentoring_international
      return 0
    end

    if p1.is_international && !p2.is_international
      return 1
    end

    if !p1.is_international && p2.is_international
      return -1
    end

    return 0

  end

  sig do
    params(
      target_rank: Integer,
      p1_rank: Integer,
      p2_rank: Integer
    ).returns(Integer)
  end
  private_class_method def self.compare_rank(target_rank:, p1_rank:, p2_rank:)
    p1_rank_difference = (p1_rank - target_rank).abs
    p2_rank_difference = (p2_rank - target_rank).abs
    
    # Reverse the comparison, since we prefer a lower rank difference (e.g. closer two ranks/seniorities).
    p2_rank_difference <=> p1_rank_difference
  end

  sig do
    params(
      mentee: Person,
      p1: Person,
      p2: Person
    ).returns(Integer)
  end
  private_class_method def self.compare_interests(mentee, p1, p2)
    mentee.interests.intersection(p1.interests.intersection).size <=> mentee.interests.intersection(p2.interests).size
  end

  sig do
    params(
      mentor: Person2025,
      p1: Person2025,
      p2: Person2025
    ).returns(Integer)
  end
  private_class_method def self.compare_mentee_seniority_allowlist(mentor:, p1:, p2:)
    allowlist = mentor.mentee_seniority_allowlist
    if allowlist.include?(p1.seniority) && !allowlist.include?(p2.seniority)
      return 1
    end

    if !allowlist.include?(p1.seniority) && allowlist.include?(p2.seniority)
      return -1
    end

    return 0
  end


  # Performs a comparison where
  #
  # Returns 1 if a == target and b != target
  # Returns -1 if a != target and b == target
  # Returns 0 otherwise
  sig do
    params(
      target: BasicObject,
      a: BasicObject,
      b: BasicObject
    ).returns(Integer)
  end
  private_class_method def self.compare_preferring_target(target:, a:, b:)
    if a == target && b != target
      return 1
    end

    if b == target && a != target
      return -1
    end

    0
  end
end

File: src/mentor_matching/preferences.py
from __future__ import annotations

from functools import cmp_to_key

from mentor_matching.models import Person
from mentor_matching.previous_matches import PreviousMatches


class Preferences:
    @classmethod
    def compute_mentor_to_mentee_preferences(
        cls,
        mentees: list[Person],
        mentors: list[Person],
        previously_matched: set[str] | None = None,
    ) -> dict[Person, list[Person]]:
        excluded_pairs = previously_matched or set()
        preferences: dict[Person, list[Person]] = {}

        if not all(person.is_mentee for person in mentees):
            raise RuntimeError("Found non mentees!")
        if not all(person.is_mentor for person in mentors):
            raise RuntimeError("Found non mentors!")

        for mentor in mentors:
            potential_mentees = [
                other_person
                for other_person in mentees
                if other_person.is_mentee
                and mentor != other_person
                and other_person.seniority < mentor.seniority
                and PreviousMatches.pair_key(mentor.email, other_person.email) not in excluded_pairs
            ]
            def mentor_comparator(p1: Person, p2: Person) -> int:
                return cls._mentor_preference_comparator(mentor=mentor, p1=p1, p2=p2)

            preferred_mentees = sorted(
                potential_mentees,
                key=cmp_to_key(mentor_comparator),
            )
            preferences[mentor] = preferred_mentees

        return preferences

    @classmethod
    def compute_mentee_to_mentor_preferences(
        cls,
        mentees: list[Person],
        mentors: list[Person],
        previously_matched: set[str] | None = None,
    ) -> dict[Person, list[Person]]:
        excluded_pairs = previously_matched or set()
        preferences: dict[Person, list[Person]] = {}

        if not all(person.is_mentee for person in mentees):
            raise RuntimeError("Found non mentees!")
        if not all(person.is_mentor for person in mentors):
            raise RuntimeError("Found non mentors!")

        for mentee in mentees:
            potential_mentors = [
                other_person
                for other_person in mentors
                if mentee != other_person
                and other_person.seniority > mentee.seniority
                and PreviousMatches.pair_key(other_person.email, mentee.email) not in excluded_pairs
            ]
            def mentee_comparator(p1: Person, p2: Person) -> int:
                return cls._mentee_preference_comparator(mentee=mentee, p1=p1, p2=p2)

            preferred_mentors = sorted(
                potential_mentors,
                key=cmp_to_key(mentee_comparator),
            )
            preferences[mentee] = preferred_mentors

        return preferences

    @classmethod
    def _mentor_preference_comparator(cls, mentor: Person, p1: Person, p2: Person) -> int:
        comparisons = (
            cls._compare_international_preference_for_mentor(mentor=mentor, p1=p1, p2=p2),
            cls._compare_mentee_seniority_allowlist(mentor=mentor, p1=p1, p2=p2),
            cls._compare_preferring_target(target=mentor.state, a=p1.state, b=p2.state),
            cls._compare_rank(target_rank=mentor.seniority, p1_rank=p1.seniority, p2_rank=p2.seniority),
        )
        return -1 * cls._sort_by_comparison_list(comparisons)

    @classmethod
    def _mentee_preference_comparator(cls, mentee: Person, p1: Person, p2: Person) -> int:
        comparisons = (
            cls._compare_international_preference_for_mentee(mentee=mentee, p1=p1, p2=p2),
            cls._compare_preferring_target(target=mentee.state, a=p1.state, b=p2.state),
            cls._compare_rank(target_rank=mentee.seniority, p1_rank=p1.seniority, p2_rank=p2.seniority),
        )
        return -1 * cls._sort_by_comparison_list(comparisons)

    @staticmethod
    def _sort_by_comparison_list(comparisons: tuple[int, ...]) -> int:
        for comparison in comparisons:
            if comparison != 0:
                return comparison
        return 0

    @staticmethod
    def _compare_international_preference_for_mentee(mentee: Person, p1: Person, p2: Person) -> int:
        if not mentee.is_international:
            return 0
        if p1.prefers_mentoring_international and not p2.prefers_mentoring_international:
            return 1
        if not p1.prefers_mentoring_international and p2.prefers_mentoring_international:
            return -1
        return 0

    @staticmethod
    def _compare_international_preference_for_mentor(mentor: Person, p1: Person, p2: Person) -> int:
        if not mentor.prefers_mentoring_international:
            return 0
        if p1.is_international and not p2.is_international:
            return 1
        if not p1.is_international and p2.is_international:
            return -1
        return 0

    @staticmethod
    def _compare_rank(target_rank: int, p1_rank: int, p2_rank: int) -> int:
        p1_rank_difference = abs(p1_rank - target_rank)
        p2_rank_difference = abs(p2_rank - target_rank)
        if p2_rank_difference < p1_rank_difference:
            return -1
        if p2_rank_difference > p1_rank_difference:
            return 1
        return 0

    @staticmethod
    def _compare_mentee_seniority_allowlist(mentor: Person, p1: Person, p2: Person) -> int:
        allowlist = mentor.mentee_seniority_allowlist
        if p1.seniority in allowlist and p2.seniority not in allowlist:
            return 1
        if p1.seniority not in allowlist and p2.seniority in allowlist:
            return -1
        return 0

    @staticmethod
    def _compare_preferring_target(target: object, a: object, b: object) -> int:
        if a == target and b != target:
            return 1
        if a != target and b == target:
            return -1
        return 0
